[쪼밀리] API 테스트/문서자동화 미션 제출합니다.

chomily opened the pull request:
안녕하세요 홍시, 쪼밀리입니다!
리뷰 잘 부탁드려요😊

JinJuMoon committed:
feat: 문서 자동화 실습 완료

JinJuMoon committed:
feat: Basic HTTP Authentication 실습 완료

JinJuMoon committed:
feat: FormAuth 및 세션을 활용한 인증 실습 완료

JinJuMoon committed:
feat: Bearer HTTP Authentication, JWT방식 실습 완료

JinJuMoon committed:
docs: 미션 1단계 요구사항을 README.md에 작성

JinJuMoon committed:
refactor: service 기능과 admin 기능을 구분하도록 URI 수정

JinJuMoon committed:
fix: LineControllerTest 실행 시 로드하는 빈 수정

JinJuMoon committed:
test: 로그인 기능에 대한 인수테스트 작성

- LoginMemberAcceptanceTest: 인수테스트 작성
- AcceptanceTest: 경로 변경

JinJuMoon committed:
test: LoginMemberController 단위테스트 작성 및 문서 자동화 적용

- LoginMemberController: 단위테스트 작성 문서 자동화 적용
- LoginMemberDocumentation: 문서에 기록할 내용 추가

JinJuMoon committed:
chore: 인수테스트 관련 커밋에서 누락된 파일을 추가로 커밋, UTF-8 인코딩 설정

JinJuMoon committed:
feat: 로그인 관련 예외를 처리하는 LoginMemberControllerAdvice 생성

JinJuMoon committed:
feat: 사용자의 회원 관리 기능 구현

- LoginMemberController 분리(MemberController, OAuthController) 및 URI 수정
- MemberDocumentation에서 OAuthDocumentation 분리
- 기존의 MemberController를 AdminMemberController로 수정
- 회원 관리 기능 인수테스트, MemberController 단위 테스트 작성
- URI변경에 따른 인터셉터 연결 설정 변경
- BearerAuthInterceptor에서 인증이 불필요한 경우에 대해 분기 처리
- MemberService에 사용자의 회원 관리 기능 추가 및 기존 메서드 네이밍 변경
- 모든 인수테스트의 클라이언트를 AcceptanceTest로 이동
- AcceptanceTest 포맷팅 수정

JinJuMoon committed:
feat: 회원정보를 조회/수정/삭제하는 페이지 연동

JinJuMoon committed:
feat: 로그인 시 서버에 로그인 요청 및 토큰 저장 구현

- 초기 데이터 생성을 위한 data.sql 작성
- README.md 업데이트

JinJuMoon committed:
feat: 토큰을 이용한 내 정보 보기, 수정, 삭제 구현

JinJuMoon committed:
feat: 회원가입 구현

- 에러 메시지 상수화
- 함수 네이밍 수정

JinJuMoon committed:
docs: 미션 2단계 요구사항을 README.md에 작성

JinJuMoon committed:
feat: 네비게이션, 인덱스 페이지에 즐겨찾기 링크 추가

JinJuMoon committed:
test: 즐겨찾기 기능 인수테스트 작성

- FavoritesResponse DTO 작성

chomily committed:
feat: 즐겨찾기 추가 기능 구현

- DB에 Favorite 테이블 추가
- FavoriteController, FavoriteService 추가
- Favorite, FavoriteRequest, FavoriteResponse 작성
- Member에 List<Favorite>추가로 Member와 Favorite의 연관 관계 적용
- 즐겨찾기 추가 시, 회원 인증을 위해 BearerAuthInterceptor에 URI추가
- Station의 생성시, datetime 디폴트 값 적용되도록 DB 스키마 수정
- 불필요한 data.sql 파일 삭제
- truncate.sql 업데이트

chomily committed:
feat: 즐겨찾기 목록 조회 기능 구현

- 즐겨찾기 목록 조회를 위한 로직 추가(FavoriteController, FavoriteService, FavoriteResponse, Member)
- 즐겨찾기 목록 조회의 Controller, Service 단위 테스트 작성
- FavoriteAcceptanceTest에 잘못된 테스트 수정

chomily committed:
feat: 즐겨찾기 삭제 기능 구현

- 즐겨찾기 삭제를 위한 로직 추가(FavoriteController, FavoriteService, Member, Favorite)
- 즐겨찾기 삭제의 Controller, Service 단위 테스트 작성
- Intercepter 경로 수정

chomily committed:
feat: (프론트엔드) 즐겨찾기 추가, 목록 조회, 삭제 기능 구현

- 즐겨찾기 id 제공을 위한 FavoriteResponse 수정
- FavoriteController 메서드 네이밍 수정
- 테스트 코드 오류 수정

chomily committed:
feat: 문서화를 위한 테스트 작성 및 문서 자동화 적용

- Favorite API 문서화를 위한 FavoriteDocumentation 작성
- API문서 포멧팅을 위한 api-guide.adoc 수정
- 문서화를 위한 테스트 코드 수정 및 Station 생성자 추가
- FavoriteControllerTest에 누락된 토큰 정보 추가

chomily committed:
docs: 예상되는 side case 추가

chomily committed:
feat: (예외 처리 추가) null 파라미터로 즐겨찾기 하는 경우, 이미 존재하는 즐겨찾기를 추가하는 경우

- Favorite, FavoriteRequest에 유효성 검사를 위해 @NotNull 추가
- Favorite의 생성자에 Null 체크 로직 추가
- FavoriteController.createFavorite()에 FavoriteRequest인자 유효성 검사를 위한 @Validated 추가
- FavoriteRequest에 toFavorite 메서드 작성
- Favorite 생성자의 중복 제거
- 즐겨찾기 중복 요청에 대한 예외 처리를 위해 DuplicateFavoriteException 작성
- Member.addFavorite에 이미 존재하는 즐겨찾기인 경우 DuplicateFavoriteException 발생 로직 추가
- Member.findFavorite의 리턴값 Optional로 수정
- FavoriteControllerTest, FavoriteServiceTest, FavoriteTest, MemberTest 단위 테스트 추가

chomily committed:
feat: (예외 처리 추가) 출발역과 도착역이 같거나 존재하지 않는 역으로 즐겨찾기를 추가하는 경우

- Favorite 생성자에 출발역과 도착역이 같지 않은지 유효성 체크 로직 추가
- InvalidStationIdException 작성
- 존재하지 않는 역으로 즐겨찾기 추가하는 경우 FavoriteService에서 InvalidStationIdException 발생하는 로직 추가
- FavoriteTest, FavoriteServiceTest에 예외 테스트 추가

chomily committed:
test: (예외 처리 테스트 코드 추가) 존재하지 않는 즐겨찾기를 삭제하는 경우

- MemberTest, FavoriteServiceTest에 예외 테스트 추가

chomily committed:
feat: 예외 처리를 위한 FavoriteControllerAdvice, MemberControllerAdvice, GlobalControllerAdvice 작성

- InvalidAuthenticationException 패키지 이동

chomily committed:
feat: 프론트엔드에 즐겨찾기 추가 버튼 색깔 토글 적용

- Member가 가진 Favorites의 StationId찾는 메서드 네이밍 수정

chomily committed:
feat: (예외 처리) null, 빈 값, 적절하지 않은 형식의 이메일로 로그인, 회원 가입, 정보 수정하는 경우

- Member, LoginRequest, MemberRequest, UpdateMemberRequest의 멤버 변수에 @NotBlank, @Email 적용
- MemberController, OAuthController의 회원가입, 로그인, 회원 정보 수정 메서드 파라미터에 @Valid 적용
- MemberControllerTest, OAuthControllerTest에 예외 테스트 추가

chomily committed:
test: MemberServiceTest에 단위 테스트 추가

hongsii commented on src/main/java/wooteco/subway/domain/favorite/Favorite.java:
```diff
@@ -0,0 +1,52 @@
+    @NotNull
+    Long preStation;
+    @NotNull
+    Long station;
```
코드를 사용하다 보면 네이밍만 보고 Station 타입이라고 오해를 불러일으킬 수 있을 듯 합니다. 명확하게 id를 뒤에 붙이는 건 어떨까요?

hongsii commented on src/main/java/wooteco/subway/service/favorite/FavoriteService.java:
```diff
@@ -0,0 +1,66 @@
+        return stations.stream()
+            .filter(station -> station.getId().equals(stationId))
+            .findFirst()
+            .orElseThrow(() -> new RuntimeException("해당하는 역이 존재하지 않습니다."));
```
위에 `InvalidStationIdException` 을 사용 중이니 해당 예외를 사용해도 좋을 것 같아요

hongsii commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -9,43 +10,55 @@
+
+	public String createToken(LoginRequest param) {
+		Member member = memberRepository.findByEmail(param.getEmail())
+			.orElseThrow(() -> new RuntimeException("해당 이메일이 존재하지 않습니다."));
```
RuntimeException 보다는 IllegalArgumentException 과 같이 구체적인 성격을 나타낼 수 있는 예외를 사용하면 더 좋을 것 같아요!
그리고 `findMemberByEmail(..)`을 재사용하면 어떨까요?

hongsii commented on src/main/java/wooteco/subway/web/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,27 @@
+        return ResponseEntity.badRequest().body(e.getMessage());
+    }
+
+    @ExceptionHandler(Exception.class)
```
4xx, 5xx 코드가 의미하는 것은 무엇일까요? 그리고 전체 예외를 핸들링하는 곳에선 어떤 코드로 처리하는 것이 적절할까요?
이 곳에서 처리되는 로직은 어떤 것일까요? 를 고민해보시면 어떨까요?

hongsii commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthInterceptor.java:
```diff
@@ -22,13 +23,14 @@ public BearerAuthInterceptor(AuthorizationExtractor authExtractor, JwtTokenProvi
+            return true;
+        }
+
+        String extractedAuth = authExtractor.extract(request, "Bearer");
```
문자열은 상수로 관리해보면 어떨까요?

hongsii commented on src/main/resources/templates/admin/index.html:
```diff
@@ -30,7 +30,7 @@
           <li class="inline-block p-1">
-            <a href="/lines" class="underline">노선 관리</a>
+              <a class="underline" href="/admin/lines">노선 관리</a>
           </li>
```
역 관리와 구간 관리도 admin 주소로 바뀌어야 하지 않을까요?

hongsii commented on src/test/java/wooteco/subway/doc/FavoriteDocumentation.java:
```diff
@@ -0,0 +1,63 @@
+public class FavoriteDocumentation {
+    public static RestDocumentationResultHandler createFavorite() {
+        return document("favorites/create",
+            requestHeaders(
+                headerWithName("authorization").description("The user's authorized token")
+            ),
```
문서에서 반복해서 쓰일 헤더는 재사용해보면 어떨까요?

hongsii commented on src/test/java/wooteco/subway/service/favorite/FavoriteServiceTest.java:
```diff
@@ -0,0 +1,204 @@
+
+        List<FavoriteResponse> favorites = favoriteService.getFavorites(member);
+
+        // Todo: 스트림으로인해 순서를 알 수 없는 경우 명확하게 테스트 할 수 있는 방법?
```
assertj에서 컬렉션 검증문을 제공하니 이를 활용해보면 어떨까요?

* https://github.com/joel-costigliola/assertj-examples/blob/master/assertions-examples/src/test/java/org/assertj/examples/IterableAssertionsExamples.java

chomily replied:
우와, 너무 유용한 자료에요! 감사감사합니다!😊
extracting과 containsOnly를 사용해서 테스트 코드 수정했습니다 : )

hongsii reviewed (changes requested):
안녕하세요 쪼밀리!
전반적으로 요구사항을 잘 구현해주셨네요 👍 
몇 가지 피드백 드렸으니, 확인하셔서 반영해주세요~

chomily committed:
refactor: 1차 피드백 반영

- Favorite 의 멤버 변수 이름 수정
- IllegalLoginRequestException 작성
- FavoriteService.mapStationById 메서드의 예외 상황에 발생시키는 예외 타입 변경
- MemberService 메서드들의 매개 변수 이름 수정
- MemberService 에서 ID, 비밀번호에 해당하는 Member를 찾을 수 없는 경우 IllegalLoginRequestException 발생
- MemberService.createToken 메서드 내부에서 findMemberByEmail메서드를 활용하도록 수정
- html에 잘못된 역관리, 구간 관리 uri 수정
- MemberDocumentation, FavoriteDocumentation에서 공통으로 사용되는 Authorization Header를 MemberDocumentation 클래스의 상수로 관리
- web.member 하위 패키지를 역할에 따라 분리(auth, interceptor, resolver)
- BearerAuthInterceptor 내부 하드 코딩 상수 값 추출
- extracting, containsOnly 를 활용해 애매모호한 테스트 명확히 수정
- 디폴트 예외 핸들러의 응답 코드 변경
- ControllerAdvice의 우선순위 지정

hongsii commented on src/main/java/wooteco/subway/domain/member/Member.java:
```diff
@@ -26,6 +46,40 @@ public Member(Long id, String email, String name, String password) {
+
+    public void addFavorite(Favorite favorite) {
+        Objects.requireNonNull(favorite);
+        if (findFavorite(favorite).isPresent()) {
```
`Collection` 을 활용하기 위해 `equals(...)` 를 구현해도 괜찮을 것 같아요! 
지금도 괜찮은 구조니 이 피드백은 참고만 해주세요~

chomily replied:
맞아요! 처음에는 페어랑 `equals()`를 구현하려고 했었어요~😊

여기서는 `preStationId`와 `stationsId`가 동일하면 같은 즐겨찾기라고 판단해야하는데, 실제 `Favorite` 클래스는 `id` 멤버변수도 가지고 있어요. 이 때 `prestationId`, `stationId`를 기준으로 `equals()`를 작성하면 다른 곳에서 API를 사용할 때 예상치 않은 결과를 가지고 올 수 있을 것이라고 생각했어요!

그리고, 현재 `Favorite`은 Entity인데 `equals()`를 구현해야한다면 `id`를 기준으로 구현해야하지 않나 싶기도 하구요...!

hongsii commented on src/main/java/wooteco/subway/domain/station/Station.java:
```diff
@@ -1,14 +1,21 @@
     private Long id;
     private String name;
+    @JsonDeserialize(using = LocalDateTimeDeserializer.class)
+    @JsonSerialize(using = LocalDateTimeSerializer.class)
```
도메인과 클라이언트 모델의 구현은 분리하는 것이 좋습니다.
그 이유는 클라이언트(혹은 API)마다 요구사항 구현이 다를 수 있으며, 클라이언트가 변경되면 도메인도 변경이 되기 때문에 변경에 취약한 구조가 됩니다.
클라이언트에서 사용할 모델을 따로 만들어보면 어떨까요?

hongsii commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthInterceptor.java:
```diff
@@ -22,13 +24,14 @@ public BearerAuthInterceptor(AuthorizationExtractor authExtractor, JwtTokenProvi
-        // TODO: 추출한 토큰값에서 email 정보 추출 (jwtTokenProvider.getSubject() 메서드 활용)
-        String email = "";
-
+        // Todo: 리팩토링 필요
+        if ("/members".equals(request.getRequestURI()) && request.getMethod().equals("POST")) {
+            return true;
+        }
```
아직 TODO가 남았네요! 리팩토링을 진행해보시면 어떨까요?

hongsii commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthInterceptor.java:
```diff
@@ -22,13 +24,14 @@ public BearerAuthInterceptor(AuthorizationExtractor authExtractor, JwtTokenProvi
+        String extractedAuth = authExtractor.extract(request, BEARER);
+        jwtTokenProvider.validateToken(extractedAuth);
+        String email = jwtTokenProvider.getSubject(extractedAuth);
         request.setAttribute("loginMemberEmail", email);
```
여기도 상수로 선언할 수 있지 않을까요?

hongsii reviewed (changes requested):
우선 피드백이 늦어서 죄송합니다 ㅠㅠ
이전 피드백 반영을 깔끔하게 잘 해주셨네요! 👍 
간단한 피드백 남겼으니, 여기까지 반영해주세요 :)

chomily committed:
refactor: 2차 피드백 반영

- 도메인이 클라이언트에게 노출되지 않도록 FavoriteResponse가 StationResponse를 멤버로 가지도록 수정
- 응답을 위한 StationResponse 멤버 변수 타입 수정과 그에 따른 생성자 수정
- 사용자 인증 관련 인터셉터 처리를 위해 커스텀 어노테이션 @RequireAuth 생성
- BearerAuthInterceptor에서 사용자 인증이 필요한 요청을 구분하는 로직 리팩토링
- 전체 변경에 따른 테스트 코드 수정

hongsii commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthInterceptor.java:
```diff
@@ -1,31 +1,37 @@
+        HttpServletResponse response, Object handler) {
+        HandlerMethod handlerMethod = (HandlerMethod)handler;
+
+        if (Objects.isNull(handlerMethod.getMethodAnnotation(RequireAuth.class))) {
```
어노테이션 활용 👍

hongsii reviewed (approved):
피드백 반영을 깔끔하게 잘 해주셨네요! 👍 
미션 진행하시느라 고생많으셨습니다! 머지할게요 :)
